pkg/controller/csr/csr_controller.go

```diff
@@ -0,0 +1,321 @@
+	csrList := &certificatesv1.CertificateSigningRequestList{}
+
+	// Filter out unnecessary CSRs. (Calico CSRs are guaranteed to have this label).
+	requirement, err := labels.NewRequirement(LabelName, selection.Exists, []string{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if err := r.client.List(ctx, csrList, &client.ListOptions{LabelSelector: labels.NewSelector().Add(*requirement)}); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	instance := &operatorv1.Installation{}
+	if err := r.client.Get(ctx, utils.DefaultInstanceKey, instance); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	certificateManager, err := certificatemanager.Create(r.client, &instance.Spec, r.clusterDomain, common.OperatorNamespace(), certificatemanager.WithLogger(reqLogger))
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	for _, csr := range csrList.Items {
+		if csr.Spec.SignerName != utils.OperatorCSRSignerName || csr.Status.Certificate != nil || !relevantCSR(&csr) {
+			// Not for us, or already signed.
+			continue
+		}
+		reqLogger.Info("Inspecting CSR with name : %v.", csr.Name)
+		certificateTemplate, err := r.validate(ctx, &csr)
+		if err != nil {
+			if errors.As(err, &systemError{}) {
+				// Validation failed due to an error unrelated to the CSR itself.
+				return reconcile.Result{}, err
+			}
+
+			csr.Status.Conditions = []certificatesv1.CertificateSigningRequestCondition{
+				{
+					Type:    certificatesv1.CertificateDenied,
+					Message: err.Error(),
+					Reason:  err.Error(),
+					Status:  corev1.ConditionTrue,
+				},
+			}
+			reqLogger.Error(err, "Rejecting the CSR.")
+			_, err = r.certificatesClient.CertificateSigningRequests().UpdateStatus(ctx, &csr, metav1.UpdateOptions{})
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			continue
+		}
+		csr.Status.Conditions = []certificatesv1.CertificateSigningRequestCondition{
+			{
+				Type:    certificatesv1.CertificateApproved,
+				Message: "Approved",
+				Reason:  "Approved",
+				Status:  corev1.ConditionTrue,
+			},
+		}
+		reqs := r.certificatesClient.CertificateSigningRequests()
+		updatedCSR, err := reqs.UpdateApproval(ctx, csr.Name, &csr, metav1.UpdateOptions{})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		reqLogger.Info("Approved CSR with name : %v.", csr.Name)
+
+		certificatePEM, err := certificateManager.SignCertificate(certificateTemplate, certificateTemplate.PublicKey)
+		if err != nil {
+			reqLogger.Error(err, "error signing certificate request")
+			return reconcile.Result{}, err
+		}
+		updatedCSR.Status.Certificate = certificatePEM
+
+		_, err = r.certificatesClient.CertificateSigningRequests().UpdateStatus(ctx, updatedCSR, metav1.UpdateOptions{})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		reqLogger.Info("Signed CSR with name : %v.", csr.Name)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+type systemError struct {
```

Comment:
This is a fairly generic name for an error type - think we need a big comment explaining what this is for and/or a rename to make it more clear when code authors should consider using this vs. a generic error. 

My understanding is that this is meant to allow callers of `validate` to determine if there was a validation error or a generic error when talking to the apiserver.

I'd suggest either:
- Not performing any API calls in the `validate` function, removing the need for this altogether. That seems the nicest conceptually, since "validate" should really just be validating a resource, not generating API calls.
- Switching this to "InvalidCSRError" which is pretty tightly scoped conceptually in comparison.

There's also an option to use the existing APIStatus error type: https://github.com/kubernetes/apimachinery/blob/master/pkg/api/errors/errors.go#L39-L43 

But I can't say for sure that covers exactly what this does...

Reply:
I opted for `invalidCertificateRequestError`. It is indeed used to differentiate between errors that lead to denying a CSR vs errors that should end the reconcile loop, due to an unexpected issue.

Motivated by:
- Fetching the pods also requires validation.
- Preferably we first do the "cheap checks" that require no api calls. 
- Keeping all validation related logic contained in this function allows us to unit test all of it comprehensively, while keeping the code relatively readable.